<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /home/tkakusa/catkin_ws/src/rover_research/urdfs/robot.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="myrobot" xmlns:xacro="http://ros.org/wiki/xacro">
  <!-- Robot base -->
  <link name="base_link">
    <visual>
      <geometry>
        <box size="0.5 0.25 0.1"/>
      </geometry>
    </visual>
    <inertial>
      <mass value="40"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <collision>
      <geometry>
        <box size="0.5 0.25 0.1"/>
      </geometry>
    </collision>
  </link>
  <link name="wheel_front_left">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </visual>
    <inertial>
      <mass value="2"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <collision>
      <geometry>
        <cylinder length="0.04" radius=".07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </collision>
  </link>
  <link name="wheel_front_right">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </visual>
    <inertial>
      <mass value="2"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <collision>
      <geometry>
        <cylinder length="0.04" radius=".07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </collision>
  </link>
  <link name="wheel_back_left">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </visual>
    <inertial>
      <mass value="2"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <collision>
      <geometry>
        <cylinder length="0.04" radius=".07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </collision>
  </link>
  <link name="wheel_back_right">
    <visual>
      <geometry>
        <cylinder length="0.04" radius="0.07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </visual>
    <inertial>
      <mass value="2"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <collision>
      <geometry>
        <cylinder length="0.04" radius=".07"/>
      </geometry>
      <origin rpy="1.57075 0 0"/>
    </collision>
  </link>
  <!-- Joints for the four wheels -->
  <joint name="base_to_front_right_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_front_right"/>
    <origin xyz="0.125 0.125 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <joint name="base_to_front_left_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_front_left"/>
    <origin xyz="0.125 -0.125 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <joint name="base_to_back_right_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_back_right"/>
    <origin xyz="-0.125 0.125 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <joint name="base_to_back_left_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="wheel_back_left"/>
    <origin xyz="-0.125 -0.125 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!--
    <gazebo>
    <plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>0.0</updateRate>
      <leftJoint>base_to_back_left_wheel</leftJoint>
      <rightJoint>base_to_back_right_wheel</rightJoint>
      <wheelSeparation>${wheelsep}</wheelSeparation>
      <wheelDiameter>${wheeldiam}</wheelDiameter>
      <torque>100</torque>
      <commandTopic>cmd_vel</commandTopic>
      <odometryTopic>odom</odometryTopic>
      <odometryFrame>odom</odometryFrame>
      <robotBaseFrame>base_link</robotBaseFrame>
      <legacyMode>false</legacyMode>
    </plugin>
  </gazebo>
-->
  <gazebo>
    <plugin filename="libgazebo_ros_skid_steer_drive.so" name="skid_steer_drive_controller">
      <updateRate>0.0</updateRate>
      <robotNamespace>/</robotNamespace>
      <leftFrontJoint>base_to_front_left_wheel</leftFrontJoint>
      <rightFrontJoint>base_to_front_right_wheel</rightFrontJoint>
      <leftRearJoint>base_to_back_left_wheel</leftRearJoint>
      <rightRearJoint>base_to_back_right_wheel</rightRearJoint>
      <wheelSeparation>0.25</wheelSeparation>
      <wheelDiameter>0.14</wheelDiameter>
      <robotBaseFrame>base_link</robotBaseFrame>
      <torque>10</torque>
      <topicName>cmd_vel</topicName>
      <broadcastTF>false</broadcastTF>
    </plugin>
  </gazebo>
  <!-- Camera Joint -->
  <joint name="camera_joint" type="fixed">
    <axis xyz="0 1 0"/>
    <origin rpy="0 0 0" xyz="0.05 0 0.02"/>
    <parent link="pan_tilt_link"/>
    <child link="camera_link"/>
  </joint>
  <!-- Camera -->
  <link name="camera_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
      <material name="red"/>
    </visual>
    <inertial>
      <mass value="1e-5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6"/>
    </inertial>
  </link>
  <!-- camera module-->
  <gazebo reference="camera_link">
    <sensor name="camera1" type="camera">
      <update_rate>0.0</update_rate>
      <camera name="head">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>800</width>
          <height>800</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>300</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
               That pixel's noise value is added to each of its color
               channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin filename="libgazebo_ros_camera.so" name="camera_controller">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>camera1</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera_link</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>
  <!-- Pan tilt controller -->
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <robotNamespace>/robot</robotNamespace>
    </plugin>
  </gazebo>
  <!-- Pan tilt link -->
  <link name="pan_tilt_link">
    <visual>
      <geometry>
        <cylinder length="0.06" radius="0.025"/>
      </geometry>
      <origin xyz="0 0 0.03"/>
    </visual>
    <inertial>
      <mass value="1"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
    <collision>
      <geometry>
        <cylinder length="0.06" radius="0.025"/>
      </geometry>
      <origin xyz="0 0 0.03"/>
    </collision>
  </link>
  <!-- Pan tilt joint -->
  <joint name="base_to_pan_tilt" type="revolute">
    <parent link="base_link"/>
    <child link="pan_tilt_link"/>
    <origin xyz="0.225 0 0.05"/>
    <axis xyz="0 0 1"/>
    <limit effort="1000.0" lower="-1.57079632679" upper="1.57079632679" velocity="0.5"/>
  </joint>
  <transmission name="base_to_pan_tilt_simple_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="base_to_pan_tilt">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>50</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="base_to_front_right_wheel_simple_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="base_to_front_right_wheel">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>50</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="base_to_front_left_wheel_simple_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="base_to_front_left_wheel">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>50</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="base_to_back_right_wheel_simple_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="base_to_back_right_wheel">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>50</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <transmission name="base_to_back_left_wheel_simple_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="base_to_back_left_wheel">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <mechanicalReduction>50</mechanicalReduction>
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </actuator>
  </transmission>
  <link name="front_ultrasonic_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </visual>
    <inertial>
      <mass value="0.001"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="front_ultrasonic_joint" type="fixed">
    <parent link="base_link"/>
    <child link="front_ultrasonic_link"/>
    <origin xyz="0.25 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <gazebo reference="front_ultrasonic_link">
    <sensor name="front_ultrasonic" type="ray">
      <always_on>true</always_on>
      <update_rate>0.0</update_rate>
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <ray>
        <scan>
          <horizontal>
            <samples>720</samples>
            <resolution>1</resolution>
            <min_angle>-0.261799</min_angle>
            <max_angle>0.261799</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.1</min>
          <max>4</max>
          <resolution>0.01</resolution>
        </range>
      </ray>
      <plugin filename="libhector_gazebo_ros_sonar.so" name="gazebo_front_ultrasonic_controller">
        <gaussianNoise>0.005</gaussianNoise>
        <topicName>front_ultrasonic</topicName>
        <frameId>front_ultrasonic_link</frameId>
      </plugin>
    </sensor>
  </gazebo>
  <link name="back_ultrasonic_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </visual>
    <inertial>
      <mass value="0.001"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="back_ultrasonic_joint" type="fixed">
    <parent link="base_link"/>
    <child link="back_ultrasonic_link"/>
    <origin xyz="-0.25 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  <gazebo reference="back_ultrasonic_link">
    <sensor name="back_ultrasonic" type="ray">
      <always_on>true</always_on>
      <update_rate>0.0</update_rate>
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <ray>
        <scan>
          <horizontal>
            <samples>720</samples>
            <resolution>1</resolution>
            <min_angle>2.87979365359</min_angle>
            <max_angle>3.40339165359</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.1</min>
          <max>4</max>
          <resolution>0.01</resolution>
        </range>
      </ray>
      <plugin filename="libhector_gazebo_ros_sonar.so" name="gazebo_back_ultrasonic_controller">
        <gaussianNoise>0.005</gaussianNoise>
        <topicName>back_ultrasonic</topicName>
        <frameId>back_ultrasonic_link</frameId>
      </plugin>
    </sensor>
  </gazebo>
  <link name="pan_ultrasonic_link">
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.02 0.02 0.02"/>
      </geometry>
    </visual>
    <inertial>
      <mass value="0.001"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <joint name="pan_ultrasonic_joint" type="fixed">
    <parent link="pan_tilt_link"/>
    <child link="pan_ultrasonic_link"/>
    <origin xyz="0.025 0 0.05"/>
    <axis xyz="0 1 0"/>
  </joint>
  <gazebo reference="pan_ultrasonic_link">
    <sensor name="pan_ultrasonic" type="ray">
      <always_on>true</always_on>
      <update_rate>0.0</update_rate>
      <pose>0 0 0 0 0 0</pose>
      <visualize>true</visualize>
      <ray>
        <scan>
          <horizontal>
            <samples>720</samples>
            <resolution>1</resolution>
            <min_angle>-0.261799</min_angle>
            <max_angle>0.261799</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.1</min>
          <max>4</max>
          <resolution>0.01</resolution>
        </range>
      </ray>
      <plugin filename="libhector_gazebo_ros_sonar.so" name="gazebo_pan_ultrasonic_controller">
        <gaussianNoise>0.005</gaussianNoise>
        <topicName>pan_ultrasonic</topicName>
        <frameId>pan_ultrasonic_link</frameId>
      </plugin>
    </sensor>
  </gazebo>
  <!-- IMU Link -->
  <link name="imu_link">
    <inertial>
      <mass value="1"/>
      <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0"/>
    </inertial>
  </link>
  <!-- Joint for the IMU -->
  <joint name="imu_joint" type="fixed">
    <parent link="base_link"/>
    <child link="imu_link"/>
  </joint>
  <gazebo>
    <plugin filename="libgazebo_ros_imu.so" name="imu_plugin">
      <alwaysOn>true</alwaysOn>
      <bodyName>imu_link</bodyName>
      <topicName>imu</topicName>
      <serviceName>imu_service</serviceName>
      <gaussianNoise>0.0</gaussianNoise>
      <updateRate>0.0</updateRate>
    </plugin>
  </gazebo>
</robot>

